lib/new-statemachine.fnl

```diff
@@ -0,0 +1,215 @@
+  [fsm current-state action]
+  (. fsm.states current-state action))
+
+(fn get-state
+  [fsm]
+  (atom.deref fsm.state))
+
+(fn signal
+  [fsm action extra]
+  "Based on the action and the fsm's current-state, set the new state and call
+  all subscribers with the previous state, new state, action, and extra"
+  (let [state (get-state fsm)
+        {: current-state : context} state]
+    (if-let [tx-fn (get-transition-function fsm current-state action)]
+            (let [
+                  ; TODO: Should we pass the whole state (current state and context) or just context?
+                  transition (tx-fn state action extra)
+                  new-state transition.state
+                  effect transition.effect]
+                  ; If next-state is nil, error: Means the action is not expected in this state
+
+              (update-state fsm new-state)
+              ; Call all subscribers
+              (each [_ sub (pairs (atom.deref fsm.subscribers))]
+                (sub {:prev-state state :next-state new-state : action : effect : extra})))
+            (if fsm.log (fsm.log.wf "Action :%s does not have a transition function in state :%s" action current-state)))))
+
+(fn subscribe
+  [fsm sub]
+  "Adds a subscriber to the provided fsm. Returns a function to unsubscribe"
+  ; Super naive: Returns a function that just removes the entry at the inserted
+  ; key, but doesn't allow the same function to subscribe more than once since
+  ; its keyed by the string of the function itself.
+  (let [sub-key (tostring sub)]
+    (atom.swap! fsm.subscribers (fn [subs sub]
+                                  (merge {sub-key sub} subs)) sub)
+    ; Return the unsub func
+    (fn []
+      (atom.swap! fsm.subscribers (fn [subs key] (tset subs key nil)) sub-key))))
+
+(fn effect-handler
+  [effect-map]
+  "
+  Takes a map of effect->function and returns a function that handles these
+  effects by calling the mapped-to function, and then calls that function's
+  return value (a cleanup function) and calls it on the next transition.
+
+  These functions must return their own cleanup function or nil.
+  "
+  ;; Create a one-time atom used to store the cleanup function
+  (let [cleanup-ref (atom.new nil)]
+    ;; Return a subscriber function
+    (fn [{: prev-state : next-state : action : effect : extra}]
+      ;; Whenever a transition occurs, call the cleanup function, if set
+      (call-when (atom.deref cleanup-ref))
+      ;; Get a new cleanup function or nil and update cleanup-ref atom
+      (atom.reset! cleanup-ref
+                   ; TODO: Should we provide everything e.g. prev-state, action, effect?
+                   (call-when (. effect-map effect) next-state extra)))))
+
+(fn create-machine
+  [template]
+  (let [fsm  {:state (atom.new {:current-state template.state.current-state :context template.state.context})
+              :states template.states
+              ; TODO: Use something less naive for subscribers
```

Comment:
What makes this naive btw?

Reply:
More naive than your stream thing. Naive that it's keyed by the function tostring, which just happens to be unique by virtue that that stringification includes its address.